contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,631 @@
+     * 
+     * @param _setToken     The set token that is being liquidated
+     */
+    function _liquidateComponents(ISetToken _setToken) internal {
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        for (uint256 i = 0; i < positions.length; i++) {
+            address token = positions[i].component;
+            uint256 tokenBalance = IERC20(token).balanceOf(address(this));
+            (, Exchange exchange) = _getMaxTokenForExactToken(tokenBalance, token, WETH);
+            _sellToken(exchange, token, tokenBalance);
+        }
+    }
+    
+    /**
+     * Issues set tokens for an exact amount of input WETH. 
+     * Acquires set token components at the best price accross uniswap and sushiswap.
+     * Uses the acquired components to issue the set tokens.
+     * 
+     * @param _setToken         Address of the set token being issued
+     * @param _minSetReceive    Minimum amount of index to receive
+     * @return setTokenAmount   Amount of set tokens issued
+     */
+    function _issueSetTokenForExactWETH(ISetToken _setToken, uint256 _minSetReceive) internal returns(uint256) {
+        uint256 wethBalance = IERC20(WETH).balanceOf(address(this));
+        
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        
+        (uint256[] memory amountEthIn, Exchange[] memory exchanges, uint256 sumEth) = _getAmountETHForIssuance(_setToken);
+
+        uint256 setTokenAmount = _acquireComponents(positions, amountEthIn, exchanges, wethBalance, sumEth);   // acquire set token components
+        
+        require(setTokenAmount > _minSetReceive, "ExchangeIssuance: INSUFFICIENT_OUTPUT_AMOUNT");
+        
+        basicIssuanceModule.issue(_setToken, setTokenAmount, msg.sender);                            // issue token
+        
+        return setTokenAmount;
+    }
+    
+    
+    /**
+     * Handles converting the contract's full WETH balance to the output
+     * token or ether and transfers it to the msg sender.
+     *
+     * @param _isOutputETH      Converts the contract's WETH balance to ETH if set to true
+     * @param _outputToken      The token to swap the contract's WETH balance to. 
+     *                          Ignored if _isOutputETH is set to true.
+     * @param _minOutputReceive Minimum amount of output token or ether to receive. This 
+     *                          function reverts if the output is less than this. 
+     * @return                  Amount of output ether or tokens sent to msg.sender
+     */
+    function _handleRedeemOutput(
+        bool _isOutputETH,
+        address _outputToken,
+        uint256 _minOutputReceive
+    )
+        internal
+        returns (uint256)
+    {
+        if(_isOutputETH) {
+            IWETH(WETH).withdraw(IERC20(WETH).balanceOf(address(this)));
+            uint256 outputAmount = address(this).balance;
+            require(outputAmount > _minOutputReceive, "ExchangeIssuance: INSUFFICIENT_OUTPUT_AMOUNT");
```

Comment:
Save this string to a constant variable to save on gas cost for deployment and to reduce repetition.